telemetry/telemetry.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2017 Microsoft. All rights reserved.
+// MIT License
+
+package telemetry
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+)
+
+// OS Details structure.
+type OSInfo struct {
+	OSType         string
+	OSVersion      string
+	KernelVersion  string
+	OSDistribution string
+}
+
+// System Details structure.
+type SystemInfo struct {
+	MemVMTotal       uint64
+	MemVMFree        uint64
+	MemUsedByProcess uint64
+	DiskVMTotal      uint64
+	DiskVMFree       uint64
+	CPUCount         int
+}
+
+// Interface Details structure.
+type InterfaceInfo struct {
+	InterfaceType         string
+	Subnet                string
+	PrimaryCA             string
+	MAC                   string
+	Name                  string
+	SecondaryCATotalCount int
+	SecondaryCAUsedCount  int
+}
+
+// CNI Bridge Details structure.
+type BridgeInfo struct {
+	NetworkMode string
+	BridgeName  string
+}
+
+// Orchestrator Details structure.
+type OrchsestratorInfo struct {
+	OrchestratorName    string
+	OrchestratorVersion string
+}
+
+// Azure CNI Telemetry Report structure.
+type Report struct {
+	StartFlag           bool
+	Name                string
+	Version             string
+	OrchestratorDetails *OrchsestratorInfo
+	OSDetails           *OSInfo
+	SystemDetails       *SystemInfo
+	InterfaceDetails    *InterfaceInfo
+	BridgeDetails       *BridgeInfo
+	VnetAddressSpace    []string
+	ErrorMessage        string
+	Context             string
+	SubContext          string
+}
+
+// ReportManager structure.
+type ReportManager struct {
+	HostNetAgentURL string
+	IpamQueryURL    string
+	ReportType      string
+	Report          *Report
+}
+
+// GetReport retrieves orchestrator, system, OS and Interface details and create a report structure.
+func (reportMgr *ReportManager) GetReport(name string, version string) (*Report, error) {
+	var err error
+	report := &Report{}
+
+	report.Name = name
+	report.Version = version
+	report.OrchestratorDetails = GetOrchestratorDetails()
+
+	report.SystemDetails, err = GetSystemDetails()
+	if err != nil {
+		return report, err
```

Comment:
See if we can have specific strings for error inside each Info structure. We can then continue to retrieve other info.